### ISSUE TITLE
group_normalization_backward produces incorrect gradients for diff_src

### ISSUE THREAD

**Issue by moadabdou** [uxlfoundation/oneDNN#3543]
**Describe the bug**

The backward pass of `group_normalization` in oneDNN produces incorrect results for `diff_src`, even when `mean`, `variance`, `diff_dst`, `scale`, and `shift` are all correctly provided. This has been tested against both:

* PyTorch's `GroupNorm(...).backward()` output
* A manual implementation using the analytical backward formula

In both reference implementations, the computed `diff_src` matches perfectly. The oneDNN output differs significantly, especially when `scale` is not all-ones.

---

**To Reproduce**

1. Input: shape `[2, 6, 4]` (interpreted as NCH with H = 4)
2. `G = 3` groups (so group size = 2 channels per group)
3. Forward and backward performed using oneDNN v3.9
4. `group_normalization_forward` with `prop_kind::forward_training`, `use_scale_shift` flag
5. `group_normalization_backward` using same descriptors, providing:

   * `src`
   * `mean`, `variance` (from forward)
   * `scale_shift`
   * `diff_dst`
   * all memory descriptors aligned and correct
6. `diff_src` output diverges from reference implementation

---

**Expected behavior**

`diff_src` should match PyTorch’s `GroupNorm(...).backward()` output within numerical tolerance (1e-5 to 1e-6). It currently does not, even when using the same inputs, saved stats, and layout.

---

**Additional context**

* Removing `dnnl_use_global_stats` did not fix the issue
* `scale_diff` and `shift_diff` match PyTorch exactly — only `diff_src` is wrong
* The reference formula used was:

  $$
  \frac{\partial L}{\partial x} = \frac{\gamma}{\sqrt{\sigma^2 + \epsilon}} \cdot \left(
    dy - \frac{1}{M} \sum dy - \hat{x} \cdot \frac{1}{M} \sum (dy \cdot \hat{x})
  \right)
  $$

  where $M$ is the number of elements per group

---

**System Info**

* oneDNN Version: 3.8.1 (confirmed backward is implemented)
* OS: Windows 11 
* CPU: Intel Core i5 11th Gen
* Tested in both single-threaded and multithreaded modes

**Comment by vpirogov** [comment]
@moadabdou, thanks for the report. Could you please share output of your test with environment variable `ONEDNN_VERBOSE=2`?

You can also inspect group normalization [reference implementation](https://github.com/uxlfoundation/oneDNN/blob/main/src/cpu/ref_group_normalization.cpp#L153) to see whether it aligns with your expectations and Pytorch.

**Comment by moadabdou** [comment]
Hi,

Thanks for your support and for confirming the backend configuration. I’ve re-tested group normalization backward using the reference implementation from oneDNN v3.8.1 (`ref:any`) and compared it directly to PyTorch's `GroupNorm`.

The backward gradients from oneDNN and PyTorch **do not match**, even when the input, scale, bias, and diff\_dst are identical.

---

### 🔍 Mismatch Details

**oneDNN `diff_src` (first 10 values):**

```
-0.431927 0.406313 0.606064 -0.288952 -0.0206073 -0.135088 -0.0550451 1.25575 0.629246 0.166415
```

**PyTorch `x.grad` :**

```python
tensor([[[[-0.6267,  0.2582],
          [ 0.4445, -0.4465]],

         [[-0.1581, -0.2733],
          [-0.2679,  1.0698]],

         [[-0.2862, -0.6783],
          [ 0.5979,  1.8953]], ...
```

This is a significant mismatch, and not a result of rounding errors or layout misalignment.

---

### 🧪 oneDNN C++ Code Used (v3.8.1, OpenMP, AVX-512):

```cpp
#include <dnnl.hpp>
#include <vector>
#include <iostream>
#include <random>

using namespace dnnl;

#include <fstream>

void save_binary(const std::string &fname, const std::vector<float> &data) {
    std::ofstream out(fname, std::ios::binary);
    out.write(reinterpret_cast<const char*>(data.data()), data.size() * sizeof(float));
}

void run_groupnorm_forward_backward(
    engine &eng, stream &strm,
    int N, int C, int H, int W, int G,
    std::vector<float> &src_data,
    std::vector<float> &scale_data,
    std::vector<float> &shift_data,
    std::vector<float> &diff_dst_data,
    std::vector<float> &dst_data,
    std::vector<float> &mean_data,
    std::vector<float> &variance_data,
    std::vector<float> &diff_src_data,
    std::vector<float> &diff_scale_data,
    std::vector<float> &diff_shift_data
) {
    const float eps = 1e-5f;
    memory::dims src_dims = {N, C, H, W};
    memory::dims stat_dims = {N * G};
    memory::dims scale_shift_dims = {C};

    auto src_md = memory::desc(src_dims, memory::data_type::f32, memory::format_tag::nchw);
    auto scale_shift_md = memory::desc(scale_shift_dims, memory::data_type::f32, memory::format_tag::x);
    auto stat_md = memory::desc(stat_dims, memory::data_type::f32, memory::format_tag::x);

    auto src_mem = memory(src_md, eng, src_data.data());
    auto dst_mem = memory(src_md, eng, dst_data.data());
    auto scale_mem = memory(scale_shift_md, eng, scale_data.data());
    auto shift_mem = memory(scale_shift_md, eng, shift_data.data());


    auto mean_mem = memory(stat_md, eng, mean_data.data());
    auto var_mem = memory(stat_md, eng, variance_data.data());
      

    // === Forward Primitive ===
    auto fwd_pd = group_normalization_forward::primitive_desc(
        eng,
        prop_kind::forward_training,
        src_md, 
        src_md,
        G, 
        eps,
        normalization_flags::use_scale | normalization_flags::use_shift
    );
    auto fwd = group_normalization_forward(fwd_pd);

    fwd.execute(strm, {
        {DNNL_ARG_SRC, src_mem},
        {DNNL_ARG_SCALE, scale_mem},
        {DNNL_ARG_SHIFT, shift_mem},
        {DNNL_ARG_MEAN, mean_mem},
        {DNNL_ARG_VARIANCE, var_mem},
        {DNNL_ARG_DST, dst_mem}
    });
    strm.wait();



    // === Backward Primitive ===

    
    auto diff_src_mem = memory(src_md, eng, diff_src_data.data());
    auto diff_dst_mem = memory(src_md, eng, diff_dst_data.data());
    auto diff_scale_mem = memory(scale_shift_md, eng);
    auto diff_shift_mem = memory(scale_shift_md, eng);


    auto bwd_pd = group_normalization_backward::primitive_desc(
        eng,
        prop_kind::backward,
        src_md, src_md, src_md,
        G, eps,
        normalization_flags::use_scale | normalization_flags::use_shift,
        fwd_pd
    );
    


    auto bwd = group_normalization_backward(bwd_pd);

    bwd.execute(strm, {
        {DNNL_ARG_SRC, src_mem},
        {DNNL_ARG_SCALE, scale_mem},
        {DNNL_ARG_SHIFT, shift_mem},
        {DNNL_ARG_MEAN, mean_mem},
        {DNNL_ARG_VARIANCE, var_mem},
        {DNNL_ARG_DIFF_DST, diff_dst_mem},
        {DNNL_ARG_DIFF_SRC, diff_src_mem},
        {DNNL_ARG_DIFF_SCALE, diff_scale_mem},
        {DNNL_ARG_DIFF_SHIFT, diff_shift_mem}
    });
    strm.wait();


    save_binary("x.bin", src_data);
    save_binary("scale.bin", scale_data);
    save_binary("bias.bin", shift_data);
    save_binary("dy.bin", diff_dst_data);
}

int main() {
    engine eng(engine::kind::cpu, 0);
    stream strm(eng);
std:: cout<< "here";

    int N = 2, C = 4, H = 2, W = 2, G = 2;
    int numel = N * C * H * W;

    std::vector<float> src_data(numel);
    std::vector<float> dst_data(numel);
    std::vector<float> scale(C, 1.0f);
    std::vector<float> shift(C, 0.0f);
    std::vector<float> mean(N * G);
    std::vector<float> var(N * G);

    std::vector<float> diff_dst(numel);  // set to 1s for simplicity
    std::vector<float> diff_src(numel);
    std::vector<float> diff_scale(C);
    std::vector<float> diff_shift(C);

    
    std::default_random_engine rng(42); // fixed seed for reproducibility
    std::normal_distribution<float> dist(1.0f, 5.0f);

    for (int i = 0; i < numel; ++i) {
        src_data[i] = dist(rng);
    }

    std::default_random_engine ng(43); // fixed seed for reproducibility
    std::normal_distribution<float> dis(2.0f, 4.0f);

    for (int i = 0; i < numel; ++i) {
        diff_dst[i] = dis(ng);
    }

    run_groupnorm_forward_backward(eng, strm,
        N, C, H, W, G,
        src_data, scale, shift, diff_dst,
        dst_data, mean, var,
        diff_src, diff_scale, diff_shift
    );

    // Print or export to file to compare with PyTorch
    std::cout << "First few diff_src values:\n";
    for (int i = 0; i < 10 && i < diff_src.size(); ++i)
        std::cout << diff_src[i] << " ";
    std::cout << "\n";

    std::cout << "First few src values:\n";
    for (int i = 0; i < 10 && i < diff_src.size(); ++i)
        std::cout << src_data[i] << " ";
    std::cout << "\n";

    return 0;
}

```

---

### 🐍 PyTorch Code:

```python
import torch

def load_bin(name, shape):
    numel = 1
    for s in shape:
        numel *= s
    with open(name, 'rb') as f:
        data = torch.tensor(
            list(torch.FloatStorage.from_buffer(f.read(), byte_order='native')), dtype=torch.float32
        )
    return data.view(*shape)

N, C, H, W = 2, 4, 2, 2
x = load_bin("x.bin", (N, C, H, W)).requires_grad_()
scale = load_bin("scale.bin", (C,))
bias = load_bin("bias.bin", (C,))
dy = load_bin("dy.bin", (N, C, H, W))

gn = torch.nn.GroupNorm(2, C)
gn.weight = torch.nn.Parameter(scale)
gn.bias = torch.nn.Parameter(bias)

y = gn(x)
y.backward(dy)

print(x.grad)
```

---

### 🧾 oneDNN Verbose Output:

```
onednn_verbose,v1,info,cpu,runtime:OpenMP,nthr:8
onednn_verbose,v1,info,cpu,isa:Intel AVX-512 with Intel DL Boost
onednn_verbose,v1,info,gpu,runtime:none
onednn_verbose,v1,info,graph,backend,0:dnnl_backend
onednn_verbose,v1,primitive,info,template:operation,engine,primitive,implementation,prop_kind,memory_descriptors,attributes,auxiliary,problem_desc,exec_time
onednn_verbose,v1,graph,info,template:operation,engine,partition_id,partition_kind,op_names,data_formats,logical_tensors,fpmath_mode,implementation,backend,exec_time
onednn_verbose,v1,primitive,create:cache_miss,cpu,group_normalization,ncsp_gnorm:any,forward_training,src:f32::blocked:abcd::f0 dst:f32::blocked:abcd::f0,,flags:CH,g2mb2ic4ih2iw2,0.0295
onednn_verbose,v1,primitive,exec,cpu,group_normalization,ncsp_gnorm:any,forward_training,src:f32::blocked:abcd::f0 dst:f32::blocked:abcd::f0,,flags:CH,g2mb2ic4ih2iw2,0.6392
onednn_verbose,v1,primitive,create:cache_miss,cpu,group_normalization,ref:any,backward,src:f32::blocked:abcd::f0 dst:f32::blocked:abcd::f0 diff_src:f32::blocked:abcd::f0,,flags:CH,g2mb2ic4ih2iw2,0.0252
onednn_verbose,v1,primitive,exec,cpu,group_normalization,ref:any,backward,src:f32::blocked:abcd::f0 dst:f32::blocked:abcd::f0 diff_src:f32::blocked:abcd::f0,,flags:CH,g2mb2ic4ih2iw2,0.0339
```

---

### ❓ Question:

Given that both implementations are using the same data and expected to follow the same normalization math, could you confirm:

1. Whether the reference `group_normalization_bwd` is expected to match PyTorch in output?
2. If PyTorch and oneDNN use different statistical reductions (e.g. mean vs. sum, biased vs. unbiased variance)?
3. If there's an internal assumption in oneDNN regarding layout or reorder that may affect correctness?

I’d be happy to share the `.bin` files if needed. I’d really appreciate a confirmation of whether this is a known discrepancy or potentially a bug in the backward computation.

**Comment by raistefintel** [comment]
Thank you for providing the reproducers, @moadabdou! I'll investigate the issue and get back to you shortly

**Comment by vpirogov** [comment]
While @raistefintel digging into the details here're some thoughts on the topic.

> Whether the reference group_normalization_bwd is expected to match PyTorch in output?

PyTorch implementation was considered [at design phase](https://github.com/uxlfoundation/oneDNN/tree/rfcs/rfcs/20230315-instance-normalization) and implementation is expected to be consistent with PyTorch. Discrepancy you see may be a bug in reference implementation or a difference in PyTorch semantics that we did not account for.

> If PyTorch and oneDNN use different statistical reductions (e.g. mean vs. sum, biased vs. unbiased variance)?

Looks like PyTorch is using biased variance, this may be the cause of observed difference.

> If there's an internal assumption in oneDNN regarding layout or reorder that may affect correctness?

Memory layouts in oneDNN do not affect the math, only the data representation in memory. As your example is using `nchw` memory layout exactly matches what PyTorch does and reorders are not required.

**Comment by raistefintel** [comment]
I've noticed a few issues with the backward pass in the reference implementation. I'll let you know when we have a confirmed fix.
